### ISSUE TITLE
GET field count per index API

### ISSUE THREAD

**Issue by Leaf-Lin** [elastic/elasticsearch#68947]
Today, there's an API to get per index field or all fields:
```
GET <index>/_field_caps?fields=*
```
which returns:
```
{
  "indices" : [
    "my_index",
    "my_index2",
    "my_index3",
    ...
  ],
  "fields" : {
    "my_field1" : {
      "keyword" : {
        "type" : "keyword",
        "searchable" : true,
        "aggregatable" : true
      }
    },
    "my_field2" : {
      "keyword" : {
        "type" : "keyword",
        "searchable" : true,
        "aggregatable" : true
      }
    },
    ...
  }
}
```
But the output from the above doesn't have break down by index, so it's not easy to troubleshoot **which** index is having mapping explosion problem.

I can potentially loop over my indices list and running `GET <index>/field_caps?fields=*` API to extract the field length per index, but again, this is not ideal.

It would be most useful if Elasticsearch can have an API out-of-the-box to count the number of fields breakdown by the indices, expected output should look like:
```
index_name      field_count
my_index        10
my_index2       20
my_index3       30
```

This can be part of the `GET indices/stats` or `cat indices` API. 

Btw, kibana might be able to take advantage of the API instead of doing its own aggregation/counting to show index pattern field count:
<img width="1187" alt="Screen Shot 2021-02-12 at 11 42 41 am" src="https://user-images.githubusercontent.com/39002973/107717356-8ce67e00-6d27-11eb-9b89-3d4e5470955f.png">

**Comment by mbrunnert** [comment]
This feature is becoming more valuable now that the documentation recommends [sizing heap based on field counts](https://www.elastic.co/guide/en/elasticsearch/reference/current/size-your-shards.html#field-count-recommendation)

**Comment by geekpete** [comment]
Seems we have field counts with Field Usage Stats api, but it's not so consumable with some top level counts.

GET /*/_field_usage_stats

Output is per shard and recursive, so getting totals is still a manual exercise.
Potential for a feature request to add some `extended` parameter to do recursive field counts and show totals at each level.

**Comment by Hythloday-zero** [comment]
https://www.elastic.co/guide/en/kibana/master/data-views-api-get.html#data-views-api-get works to get the raw numbers similar to the Kibana UI screenshot, but it _doesn't_ get us a fields.count value which would be valuable.

**Comment by shaigbdb** [comment]
> Seems we have field counts with Field Usage Stats api, but it's not so consumable with some top level counts.
> 
> `GET /*/_field_usage_stats`
> 
This returns field usage information - I'm not seeing a field count in the output.

**Comment by geekpete** [comment]
Sorry I meant the response returns the lists of fields but counting them is something you need to do manually.

**Comment by lucabelluccini** [comment]
Especially following the introduction of `index.mapping.total_fields.ignore_dynamic_beyond_limit` via https://github.com/elastic/elasticsearch/pull/96235 (and it being used in 8.14+ on several index templates of Fleet), it is crucial to get the field count at runtime in order to eventually automate / warn if the field count is reaching the limit.

FYI @flash1293 / @felixbarny

**Comment by felixbarny** [comment]
I agree that this would be very useful. I feel like we're re-implementing this on the client side in multiple places (and not always in the correct way). For example, in the dataset quality page. cc @achyutjhunjhunwala @gbamparop.

I think exposing this in some ES API would be relatively straightforward as we already have a method for this in `MappingLookup`:

https://github.com/elastic/elasticsearch/blob/8db918110c377830ff17da49e980be8dc0e9524e/server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java#L250-L255

**Comment by VimCommando** [comment]
On a related note I would _love_ to see field counts, and breakdown by field type, in the `<index>/_stats` API. It would make alerting for mapping explosions trivial.

**Comment by felixbarny** [comment]
I've created a quick POC for adding the fields count to the index stats API: https://github.com/elastic/elasticsearch/pull/116438

**Comment by felixbarny** [comment]
I discussed this with @dakrone. A challenge with adding this to the index stats API is that this API isn't available on serverless, as it's considered to be exposing too many low-level details. However, for this use case we'll want to have a user-accessible API.

We discussed adding this to get mappings. It has some technical challenges as get mappings is answered by the coordinating node, getting the unparsed mapping from cluster state. To call `MappingLookup#getTotalFieldsCount`, we would either need to delegate to the primary shard, or parse the mapping on the coordinating node. It'll also be a bit awkward to return a property for the field count from get mappings that's not allowed to be present in a put mappings request. It may break use cases where a client gets the mapping, modifies it, and updates it. Only adding the field count in the get mapping response when a verbose flag is enabled could mitigate the breaking aspect of it. The awkwardness is still there.

**Comment by javanna** [comment]
I can see how having clear field count per index is valuable, especially as this is reverse-engineered now in multiple places, and probably not aligned with the ES logic necessarily. I see the challenges with using index stats API, as well as get mappings. It sounds like this info could be added to field caps more easily, although it does not map strictly to the capabilities of fields.

**Comment by felixbarny** [comment]
I forgot to mention that we also discussed adding this to field caps. A challenge with that approach is that field caps isn't focused on a particular index, but at an index pattern. Even if you're looking at just a single data stream, it has multiple backing indices, where each may have a different number of fields. While we could add another section in the field_caps response that lists the number of fields for each matching index, it seems to go against the spirit of that API which aggregates data for fields across multiple indices. For this, we probably want an API that's focussed on indices rather than fields.

Is the [cat indices API](https://www.elastic.co/guide/en/elasticsearch/reference/current/cat-indices.html) available on serverless?

**Comment by dakrone** [comment]
> Is the [cat indices API](https://www.elastic.co/guide/en/elasticsearch/reference/current/cat-indices.html) available on serverless?

Yes, cat indices is available for Serverless (though it should be a human-invoked thing, and not used or relied upon programmatically).

**Comment by felixbarny** [comment]
The index stats still seems like the most appropriate place to add this. It's index-centric and already fetches stats from the shards rather than just answering on the coordinating node, which makes accessing `MappingLookup#getTotalFieldsCount` straightforward. The only downside is that it doesn't work for serverless. But we could think about allowing the index stats api for serverless but limit the [index metrics](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-stats.html#index-stats-api-path-params) that we're exposing. But I'm not sure if there's precedent for partially exposing an API in serverless.

**Comment by lucabelluccini** [comment]
This is still causing a lot of pain to investigate problems with `ignore_dynamic_beyond_limit` introduced in https://github.com/elastic/kibana/pull/178398

Main reason is the field count in Kibana is way lower than the "count" done on Elasticsearch side.

**Comment by felixbarny** [comment]
> Main reason is the field count in Kibana is way lower than the "count" done on Elasticsearch side.

Is that something we can fix on the Kibana side? I agree that ideally, there should be an API in ES to expose that information but in the meantime, we could fix the bug in Kibana so that it has the same logic to count the number of fields.

**Comment by javanna** [comment]
Could we clarify why Kibana needs to replicate fields counting? I though that the idea behind using ignore_dynamic_beyond_limit was to have less total fields limit issues, as it's a graceful behaviour compared to rejecting docs.

**Comment by lucabelluccini** [comment]
The usability problem with `ignore_dynamic_beyond_limit` in general is:
- Kibana says the index has N fields, where N << limit
- ES ignores fields because the way ES counts fields is different from what is reported from Kibana. ES internally "counts" K fields, K > limit

Therefore, users and support have hard time to understand why fields started to be ignored as Kibana tells there are 740 fields in a given index/data view.

We would like to have the "real" count used by Elasticsearch which triggered ignoring fields.
This can help automation/alerts/ui hints to tell users: "this data view contains indices with ignored fields because those indices X,Y,Z exceeded the limit of dynamic fields".

---

On top of this, maybe we should file it separately, is to trigger a warning if a query makes use of a field which is ignored in 1 or more indices being searched on.

**Comment by javanna** [comment]
Maybe a silly question, but I think I am missing why Kibana needs to be aware of how many fields there are in the mappings in the first place.

**Comment by lucabelluccini** [comment]
The original ask is to have an API on Elasticsearch side to expose the count.
It would be very useful for monitoring, automated alerts, analyzers - and troubleshooting.

As users typically interact and manage their cluster through Kibana, having the count of fields in Stack Management would be another good place to leverage the count and show this info there for troubleshooting purposes (self service).
Or even the "new" `Data Set Quality` app might leverage such info.

We had already several users hit by a problem related to the introduction of `ignore_dynamic_beyond_limit` ([as it wasn't well advertised in 8.14](https://github.com/elastic/elasticsearch/issues/68947#issuecomment-2460429714)) and being unable to easily pinpoint which indices have exceeded the limit, apart having the access to the client side logs where we were getting the `Limit of total fields [10000] has been exceeded while adding...`.

Another approach would be to have a boolean flag reporting if at least 1 field was ignored (due to limit or malformed).

In short, without focusing on the implementation, I think it would be nice to know:
- How many fields do I have in this index?
- Did we hit the field count limit in this index (with or without `ignore_dynamic_beyond_limit`)?
- Are there ignored fields in this index?
    - If yes, how many due to `ignore_malformed`?
    - If yes, how many due to `ignore_dynamic_beyond_limit`?
    - If yes, which fields have been ignored?
- How many documents have `_ignored` fields in this index?

**Comment by javanna** [comment]
I get the high-level ask, that is clear to me.

I don't follow the part where Kibana gets mentioned: 

> Main reason is the field count in Kibana is way lower than the "count" done on Elasticsearch side.

where does Kibana show the field count today and what does it do with it? Pardon my ignorance there. I do get that if the field count was exposed clearly from ES there would be no need to reverse engineer it outside of ES.

**Comment by lucabelluccini** [comment]
No problem and that's great to check as we need to be sure we're picking the correct solution for the problem(s).

> where does Kibana show the field count today and what does it do with it? Pardon my ignorance there. I do get that if the field count was exposed clearly from ES there would be no need to reverse engineer it outside of ES.

Right now, Kibana shows the field count in data views, not really on specific indices 

<img width="722" alt="Image" src="https://github.com/user-attachments/assets/209f6f68-d73f-481b-b0b8-c33337b452ec" />

**Comment by maryam-saeidi** [comment]
Hi @javanna, Any update on this request? We also have a use case to monitor the number of fields in alerting indices and ensure we are not hitting the mapping limit, and if that happens, what is the magnitude of the issue? (by comparing the number of fields to the limit)

From what I read in this thread, @felixbarny already has a [PoC](https://github.com/elastic/elasticsearch/pull/116438), and the remaining question is where to expose the field count per index. Who should we talk to for making that decision?

|Option|Issue|
|---|---|
|index stats API ([comment 1](https://github.com/elastic/elasticsearch/issues/68947#issuecomment-2474200804), [comment 2](https://github.com/elastic/elasticsearch/issues/68947#issuecomment-2506092018))|Not available in Serverless (maybe it can be partially exposed in Serverless?)|
|mapping API ([comment](https://github.com/elastic/elasticsearch/issues/68947#issuecomment-2474200804))|Returning a field that will not be part of the put request, possible complication due to adding a verbose flag|
|fields cap ([comment](https://github.com/elastic/elasticsearch/issues/68947#issuecomment-2482734937))|Focuses on the index pattern instead of index|
